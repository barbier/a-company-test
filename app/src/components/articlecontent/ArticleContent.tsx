import { useContext } from 'react'
import { convertToDate } from '../../utils/util'

const { ArticleContext } = require('../../contexts/context')

const {
  Content,
  ContentAuthor,
  ContentHeader,
  ContentTitle,
  ContentPublishedDate,
  ContentImage,
  ContentText,
  ContentFooter,
  ContentSource
} = require('../../styles/styles')

const ArticleContent: React.FC = () => {
  // eslint-disable-next-line
  const [selectedArticle, setSelectedArticle] = useContext(ArticleContext)

  const {
    title,
    author,
    publishedAt,
    urlToImage,
    content,
    url,
    source,
  } = selectedArticle
  
  if (!selectedArticle.title) {
    return null
  }

  return (
    <Content>
      <ContentHeader>
        <ContentTitle>{title}</ContentTitle>
        <ContentAuthor>
          Author: 
          {
            author
            ? author
            : 'Unknown'
          }
        </ContentAuthor>
        <ContentPublishedDate>
          Published at: {convertToDate(publishedAt)}
        </ContentPublishedDate>
      </ContentHeader>
      <ContentImage src={urlToImage} />
      <ContentText>{content}</ContentText>
      <ContentFooter>
        Source: 
        <ContentSource href={url} target='_blank'>
          {source?.name}
        </ContentSource>
      </ContentFooter>
    </Content>
  )
}

export default ArticleContent