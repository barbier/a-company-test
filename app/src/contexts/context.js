import { createContext } from 'react'

export const ArticleContext = createContext()
