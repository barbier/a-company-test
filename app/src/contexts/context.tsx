import React, { useState, createContext, useContext } from "react";
import { IArticle } from "../interfaces/Article";
import { IContext } from "../interfaces/ContextProvider";

const useArticle = (): any => {
  const [selectedArticle, setSelectedArticle] = useState({});

  return {
    selectedArticle,
    chooseArticle: (article: IArticle) => setSelectedArticle(article),
  };
};

const ArticleContext = createContext({} as any);

export const ArticleContextProvider: React.FC<{ children: any }> = ({
  children,
}: IContext) => (
  <ArticleContext.Provider value={useArticle()}>
    {children}
  </ArticleContext.Provider>
);

export const useSelectedArticle = (): any =>
  useContext(ArticleContext).selectedArticle;
export const useSetSelectedArticle = (article: IArticle): any =>
  useContext(ArticleContext).chooseArticle(article);
