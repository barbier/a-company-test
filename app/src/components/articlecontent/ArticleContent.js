import { useContext } from 'react'
import { ArticleContext } from '../../contexts/context'
import { Content, ContentAuthor, ContentHeader, ContentTitle, ContentPublishedDate, ContentImage, ContentText, ContentFooter, ContentSource } from '../../styles/styles'
import { convertToDate } from '../../utils/util'

function ArticleContent() {
  // eslint-disable-next-line
  const [selectedArticle, setSelectedArticle] = useContext(ArticleContext)

  const {
    title,
    author,
    publishedAt,
    urlToImage,
    content,
    url,
    source,
  } = selectedArticle
  
  if (!selectedArticle.title) {
    return null
  }

  return (
    <Content>
      <ContentHeader>
        <ContentTitle>{title}</ContentTitle>
        <ContentAuthor>
          Author: 
          {
            author
            ? author
            : 'Unknown'
          }
        </ContentAuthor>
        <ContentPublishedDate>
          Published at: {convertToDate(publishedAt)}
        </ContentPublishedDate>
      </ContentHeader>
      <ContentImage src={urlToImage} />
      <ContentText>{content}</ContentText>
      <ContentFooter>
        Source: 
        <ContentSource href={url} target='_blank'>
          {source?.name}
        </ContentSource>
      </ContentFooter>
    </Content>
  )
}

export default ArticleContent