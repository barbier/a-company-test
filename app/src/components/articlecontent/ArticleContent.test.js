import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { ArticleContext } from '../../contexts/context'
import ArticleContent from './ArticleContent'

describe("Test Article Content rendering options", () => {
  test("It should match snapshot", () => {
    const selectedArticle = {
      "source": {
        "id": "cnbc",
        "name": "CNBC"
      },
      "author": "Amelia Lucas",
      "title": "Coca-Cola stock jumps as strong sales of Coke Zero Sugar continue to drive revenue growth - CNBC",
      "description": "Coca-Cola on Friday reported quarterly earnings that met analysts' expectations.",
      "url": "https://www.cnbc.com/2019/10/18/coca-cola-ko-earnings-q3-2019.html",
      "urlToImage": "https://image.cnbcfm.com/api/v1/image/105375236-1533317591071rtx6drjw.jpg?v=1571332846",
      "publishedAt": "2019-10-18T10:42:46Z",
      "content": "Coca-Cola on Friday reported quarterly revenue that topped analysts' expectations as more customers are drawn in by healthier options, like Zero Sugar soda and smaller size cans. \r\nShares of the company jumped 1.6% in premarket trading.\r\n\"Our performance give… [+1887 chars]"
    }
    const setSelectedArticle = () => {}
    const component = renderer.create(
      <ArticleContext.Provider value={[selectedArticle, setSelectedArticle]}>
        <ArticleContent />
      </ArticleContext.Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("It shouldn't render nothing if there's no article selected", () => {
    const selectedArticle = {}
    const setSelectedArticle = () => {}

    const {container} = render(
      <ArticleContext.Provider value={[selectedArticle, setSelectedArticle]}>
        <ArticleContent />
      </ArticleContext.Provider>
    )
    const articleTag = container.querySelector('article')

    expect(articleTag).toBeFalsy()
  })

  test("It shouldn't render normaly when an article is selected", () => {
    const selectedArticle = {
      "source": {
        "id": "cnbc",
        "name": "CNBC"
      },
      "author": "Amelia Lucas",
      "title": "Coca-Cola stock jumps as strong sales of Coke Zero Sugar continue to drive revenue growth - CNBC",
      "description": "Coca-Cola on Friday reported quarterly earnings that met analysts' expectations.",
      "url": "https://www.cnbc.com/2019/10/18/coca-cola-ko-earnings-q3-2019.html",
      "urlToImage": "https://image.cnbcfm.com/api/v1/image/105375236-1533317591071rtx6drjw.jpg?v=1571332846",
      "publishedAt": "2019-10-18T10:42:46Z",
      "content": "Coca-Cola on Friday reported quarterly revenue that topped analysts' expectations as more customers are drawn in by healthier options, like Zero Sugar soda and smaller size cans. \r\nShares of the company jumped 1.6% in premarket trading.\r\n\"Our performance give… [+1887 chars]"
    }
    const setSelectedArticle = () => {}

    const {container} = render(
      <ArticleContext.Provider value={[selectedArticle, setSelectedArticle]}>
        <ArticleContent />
      </ArticleContext.Provider>
    )
    const articleTag = container.querySelector('article')

    expect(articleTag).toBeTruthy()
  })
})