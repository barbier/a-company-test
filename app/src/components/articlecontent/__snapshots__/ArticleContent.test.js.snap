// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Article Content rendering options It should match snapshot 1`] = `
<article
  className="css-6p15e3"
>
  <header
    className="css-1etpri8"
  >
    <h1
      className="css-ybopvn"
    >
      Coca-Cola stock jumps as strong sales of Coke Zero Sugar continue to drive revenue growth - CNBC
    </h1>
    <h2
      className="css-ybopvn"
    >
      Author:
      Amelia Lucas
    </h2>
    <small
      className="css-1e2fy0l"
    >
      Published at: 
      10/18/2019
    </small>
  </header>
  <img
    className="css-1lb5oxm"
    src="https://image.cnbcfm.com/api/v1/image/105375236-1533317591071rtx6drjw.jpg?v=1571332846"
  />
  <p
    className="css-1rcv8gm"
  >
    Coca-Cola on Friday reported quarterly revenue that topped analysts' expectations as more customers are drawn in by healthier options, like Zero Sugar soda and smaller size cans. 
Shares of the company jumped 1.6% in premarket trading.
"Our performance giveâ€¦ [+1887 chars]
  </p>
  <footer
    className="css-1oteowz"
  >
    Source:
    <a
      className="css-1ors561"
      href="https://www.cnbc.com/2019/10/18/coca-cola-ko-earnings-q3-2019.html"
      target="_blank"
    >
      CNBC
    </a>
  </footer>
</article>
`;
